core/src/main/java/io/grpc/inprocess/InProcessTransport.java

```diff
@@ -442,66 +448,80 @@ public void setListener(ServerStreamListener serverStreamListener) {
       public void request(int numMessages) {
         boolean onReady = clientStream.serverRequested(numMessages);
         if (onReady) {
-          synchronized (this) {
+          synchronized (this) { // TODO How should this be handled (probably just remove)
             if (!closed) {
-              clientStreamListener.onReady();
+              syncContext.executeLater(() -> clientStreamListener.onReady());
             }
           }
         }
+        syncContext.drain();
       }
 
       // This method is the only reason we have to synchronize field accesses.
+      // Using a SynchronizationContext to avoid possibility of deadlock in direct executors.
```

Comment:
Put this comment on the syncContext itself instead of on these methods. All calls to it must be through the syncContext in this new approach.